File: czech-railway-transportation/src/main/java/cz/jalasoft/transportation/czechrailway/CzechRailwayTransportation.java
package cz.jalasoft.transportation.czechrailway;

import cz.jalasoft.net.http.HttpClient;
import cz.jalasoft.net.http.netty.NettyHttpClient;
import cz.jalasoft.transportation.*;
import cz.jalasoft.transportation.czechrailway.page.TrainDetailPage;
import cz.jalasoft.transportation.exception.TransportInfoRetrievalException;
import cz.jalasoft.transportation.exception.TransportRetrievalException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collection;

import static cz.jalasoft.util.ArgumentAssertion.mustNotBeNull;
import static cz.jalasoft.util.ArgumentAssertion.mustNotBeNullOrEmpty;

/**
 * Created by Honza Lastovicka on 25.6.15.
 */
public final class CzechRailwayTransportation implements Transportation {

    private static final Logger LOGGER = LoggerFactory.getLogger(CzechRailwayTransportation.class);

    private final Carrier carrier;

    private final TrainDetailPageFlow trainDetailFlow;


    public CzechRailwayTransportation() {
        carrier = new CzechRailwayCarrier();
        HttpClient httpClient = new NettyHttpClient();

        this.trainDetailFlow = new TrainDetailPageFlow(httpClient);
        LOGGER.debug("Czech Railway Transportation has been instantiated");
    }

    @Override
    public Collection<Transport> lookupTransport(String transport) throws TransportRetrievalException {
        mustNotBeNullOrEmpty(transport, "Name of transport");

        LOGGER.debug("Transport is being looked up: {}", transport);

        Collection<TrainDetailPage> page = trainDetailFlow.trainDetail(transport);



        throw new TransportRetrievalException(transport, "Unexpected page format. No possible state of the page has been indicated by analyzing page content.");
    }

    @Override
    public Schedule querySchedule(Transport transport) throws TransportInfoRetrievalException {
        mustNotBeNull(transport, "Transport");

        LOGGER.debug("Scheduler is being queried for transport {}", transport);
        return null;
    }

    @Override
    public Position queryPosition(Transport transport) throws TransportInfoRetrievalException {
        mustNotBeNull(transport, "Transport");

        LOGGER.debug("Position is being queried for transport {}", transport);
        return null;
    }

    @Override
    public Carrier carrier() {
        return carrier;
    }
}

File: czech-railway-transportation/src/main/scala/cz/jalasoft/transportation/czechrailway/CzechRailwayTransportation.scala
package cz.jalasoft.transportation.czechrailway

import cz.jalasoft.net.http.netty.NettyHttpClient
import cz.jalasoft.transportation._
import cz.jalasoft.transportation.czechrailway.flow.TrainDetailPageFlow
import cz.jalasoft.transportation.czechrailway.page.TrainDetailPage

import scala.collection.JavaConversions._

/**
 * Created by honzales on 28.6.15.
 */
class CzechRailwayTransportation extends Transportation with Loggable {

  debug("Czech railway transportation has been instantiated")

  private val trainPageToTransport : TrainDetailPage => Transport = {
    p => {
      Transport
        .newTransport
        .carrier(p.providerName)
        .code(p.trainCode)
        .name(p.trainName.orElse(null))
        .get()
    }
  }

  override def findTransport(transport: String): java.util.Collection[Transport] = {
    require(transport != null && !transport.isEmpty, "Train name must not be null or empty")

    debug("Transport is being looked up for " + transport)

    val client = new NettyHttpClient

    try {
      TrainDetailPageFlow(client) flow(transport) map trainPageToTransport
    } finally {
      client.close()
    }
  }

  override def querySchedule(transport: Transport): Schedule = {
    require(transport != null, "Transport must not be null")

    debug("Schedule is being queried for transport {}.", transport)

    val client = new NettyHttpClient

    try {
      val page = (TrainDetailPageFlow(client) flow(transport))
      val schedule = page.schedule()

    } finally {
      client.close()
    }
    null
  }


  override def queryPosition(transport: Transport): Position = {
    debug("Position is being queried for transport " + transport)
    null
  }

  override def carrier(): Carrier = CzechRailwayCarrier

}
